src/middlewared/middlewared/plugins/cluster_linux/ctdb_private.py

```diff
@@ -0,0 +1,232 @@
+from middlewared.schema import Dict, IPAddr, Str
+from middlewared.service import (accepts, private, job,
+                                 CRUDService, ValidationErrors,
+                                 CallError, filterable)
+from middlewared.utils import filter_list
+from middlewared.plugins.cluster_linux.utils import CTDBConfig
+import pathlib
+
+
+SHARED_VOL = CTDBConfig.CTDB_LOCAL_MOUNT.value
+PRI_IP_FILE = CTDBConfig.GM_PRI_IP_FILE.value
+ETC_IP_FILE = CTDBConfig.ETC_PRI_IP_FILE.value
+PRI_LOCK = CTDBConfig.PRI_LOCK.value
+
+
+class CtdbIpService(CRUDService):
+
+    class Config:
+        namespace = 'ctdb.private.ips'
+
+    @filterable
+    def query(self, filters=None, options=None):
+
+        ips = []
+
+        # logic is as follows:
+        #   1. if ctdb daemon is started
+        #       ctdb just reads the `ETC_IP_FILE` and loads the
+        #       ips written there into the cluster. However,
+        #       if a private ip is added/removed, it doesn't
+        #       mean the ctdb cluster has been reloaded to
+        #       see the changes in the file. So return what
+        #       the daemon sees.
+        #   2. else
+        #       if the `SHARED_VOL` is mounted and `ETC_IP_FILE` exists
+        #       and is a symlink and the symlink is pointed to the
+        #       `PRI_IP_FILE` then read it and return the contents
+        if self.middleware.call_sync('service.started', 'ctdb'):
+            ips = self.middleware.call_sync('ctdb.general.listnodes')
+            ips = list(map(lambda i: dict(i, id=i['address']), ips))
+        else:
+            try:
+                mounted = pathlib.Path(SHARED_VOL).is_mount()
+            except Exception:
+                # can happen when mounted but glusterd service
+                # is stopped/crashed etc
+                mounted = False
+
+            if mounted:
+                pri_ip_file = pathlib.Path(PRI_IP_FILE)
+                etc_ip_file = pathlib.Path(ETC_IP_FILE)
+                if pri_ip_file.exists():
+                    if etc_ip_file.is_symlink() and etc_ip_file.resolve() == pri_ip_file:
+                        with open(PRI_IP_FILE) as f:
+                            for i in f.read().splitlines():
+                                ips.append({
+                                    'id': i,
+                                    'address': i,
+                                })
+
```

Comment:
What is layout of expected dict here? `ctdb listnodes -j` should return something like this for the nodelist:
`[ { "pnn": 0, "address": "<ip address as string>"} ]` (no`id` key). If we read it from the nodes file, we add an `id` key. These should probably be consistent. I'll look up logic for how we determine `pnn` in ctdb. It might be based on order of list (skipping disabled/commented-out entries).

Reply:
I add an `id` key to what the `ctdb listnodes -j` returns as well as what we read from the private IP file. This is to make `get_instance` work in the `do_delete` method.

Reply:
```
/* Read a nodes file into a node map */
static struct ctdb_node_map *ctdb_read_nodes_file(TALLOC_CTX *mem_ctx,
                                                  const char *nlist)
{
        char **lines;
        int nlines;
        int i;
        struct ctdb_node_map *nodemap;

        nodemap = talloc_zero(mem_ctx, struct ctdb_node_map);
        if (nodemap == NULL) {
                return NULL;
        }

        lines = file_lines_load(nlist, &nlines, 0, mem_ctx);
        if (lines == NULL) {
                return NULL;
        }

        while (nlines > 0 && strcmp(lines[nlines-1], "") == 0) {
                nlines--;
        }

        for (i=0; i<nlines; i++) {
                char *node;
                uint32_t flags;
                size_t len;

                node = lines[i];
                /* strip leading spaces */
                while((*node == ' ') || (*node == '\t')) {
                        node++;
                }

                len = strlen(node);

                /* strip trailing spaces */
                while ((len > 1) &&
                       ((node[len-1] == ' ') || (node[len-1] == '\t')))
                {
                        node[len-1] = '\0';
                        len--;
                }

                if (len == 0) {
                        continue;
                }
                if (*node == '#') {
                        /* A "deleted" node is a node that is
                           commented out in the nodes file.  This is
                           used instead of removing a line, which
                           would cause subsequent nodes to change
                           their PNN. */
                        flags = NODE_FLAGS_DELETED;
                        node = discard_const("0.0.0.0");
                } else {
                        flags = 0;
                }
                if (! node_map_add(nodemap, node, flags)) {
                        talloc_free(lines);
                        TALLOC_FREE(nodemap);
                        return NULL;
                }
        }

        talloc_free(lines);
        return nodemap;
}
```
That's the logic for generating the `nodes` mapping in ctdb. It reads the `nodes` file and `pnn` is based on order in file. Disabled nodes (ones beginning with a `#` character) are present to preserve pnn value, but ip address set to 0.0.0.0 and a flag set NODE_FLAGS_DELETED (internally in ctdb).

Reply:
TL;DR. `pnn` and `id` in this case should end up being the same. Rename either the one or the other (maybe easier to use `pnn` since that's referenced in other places in ctdb output.

Reply:
Do you think it's worthwhile to add an `enabled` key to this and make it correspond to the ip string starting with `#`?

Reply:
Scratch that. I just realized that `control_listnodes()` skips entries with NODE_FLAGS_DELETED.

Reply:
Adding an `enabled` key to `ctdb listnodes -j` output would be beneficial.

Reply:
For that matter, I can add deactivated nodes to the control_listnodes() JSON output and add a boolean key `enabled` to differentiate the two if you want to add an extra key to this. Not sure if there's a benefit to allowing users to deactivate entries in the nodes file in this way.

Reply:
I've changed `id` to be an alias for `pnn` in the `query()` methods. I've also just defaulted to returning `active = False` and `available = False` if we're  not able to query the ctdb daemon directly.